Serialization bug resulting in a segfault when a gensym'ed closure in Main is used with Distributed

With combination of Dagger, FLoops, and FoldsDagger, I'm getting a strange error from the Julia compiler. I put an MWE script + project in https://gist.github.com/ca3410b16939c9715b048ea1dc528a12 which can be run by:

```console
$ git clone https://gist.github.com/ca3410b16939c9715b048ea1dc528a12 mwe
...
$ cd mwe
$ ./mwe.jl
+ julia --color=yes --startup-file=no -p1 -e 'using Pkg; Pkg.instantiate()'
+ exec julia --color=yes --startup-file=no -p1 -e 'include(popfirst!(ARGS))' ./mwe.jl
      From worker 2:    Internal error: encountered unexpected error in runtime:
      From worker 2:    UndefRefError()
      From worker 2:    istopfunction at ./compiler/utilities.jl:49 [inlined]
      From worker 2:    abstract_call_known at ./compiler/abstractinterpretation.jl:863
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:926
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:911
      From worker 2:    abstract_eval at ./compiler/abstractinterpretation.jl:1005
      From worker 2:    typeinf_local at ./compiler/abstractinterpretation.jl:1270
      From worker 2:    typeinf_nocycle at ./compiler/abstractinterpretation.jl:1326
      From worker 2:    typeinf at ./compiler/typeinfer.jl:12
...
```

The error is thrown at

https://github.com/JuliaLang/julia/blob/c0f9666d0b94b213c7ff9e64a7b4e5268aa0e18b/base/compiler/utilities.jl#L49

I get the same error in Julia 1.5.2, 1.6.0-rc1, and 1.7.0-DEV.585.

<details>
 <summary>Full traceback</summary>

```
      From worker 2:    Internal error: encountered unexpected error in runtime:
      From worker 2:    UndefRefError()
      From worker 2:    istopfunction at ./compiler/utilities.jl:49 [inlined]
      From worker 2:    abstract_call_known at ./compiler/abstractinterpretation.jl:1018
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1100
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1080
      From worker 2:    abstract_eval_statement at ./compiler/abstractinterpretation.jl:1220
      From worker 2:    typeinf_local at ./compiler/abstractinterpretation.jl:1542
      From worker 2:    typeinf_nocycle at ./compiler/abstractinterpretation.jl:1602
      From worker 2:    _typeinf at ./compiler/typeinfer.jl:214
      From worker 2:    typeinf at ./compiler/typeinfer.jl:209
      From worker 2:    typeinf_edge at ./compiler/typeinfer.jl:811
      From worker 2:    abstract_call_method at ./compiler/abstractinterpretation.jl:516
      From worker 2:    abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:142
      From worker 2:    abstract_call_known at ./compiler/abstractinterpretation.jl:1059
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1100
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1080
      From worker 2:    abstract_eval_statement at ./compiler/abstractinterpretation.jl:1220
      From worker 2:    typeinf_local at ./compiler/abstractinterpretation.jl:1542
      From worker 2:    typeinf_nocycle at ./compiler/abstractinterpretation.jl:1602
      From worker 2:    _typeinf at ./compiler/typeinfer.jl:214
      From worker 2:    typeinf at ./compiler/typeinfer.jl:209
      From worker 2:    typeinf_edge at ./compiler/typeinfer.jl:811
      From worker 2:    abstract_call_method at ./compiler/abstractinterpretation.jl:516
      From worker 2:    abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:142
      From worker 2:    abstract_call_known at ./compiler/abstractinterpretation.jl:1059
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1100
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1080
      From worker 2:    abstract_eval_statement at ./compiler/abstractinterpretation.jl:1220
      From worker 2:    typeinf_local at ./compiler/abstractinterpretation.jl:1542
      From worker 2:    typeinf_nocycle at ./compiler/abstractinterpretation.jl:1602
      From worker 2:    _typeinf at ./compiler/typeinfer.jl:214
      From worker 2:    typeinf at ./compiler/typeinfer.jl:209
      From worker 2:    typeinf_edge at ./compiler/typeinfer.jl:811
      From worker 2:    abstract_call_method at ./compiler/abstractinterpretation.jl:516
      From worker 2:    abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:142
      From worker 2:    abstract_call_known at ./compiler/abstractinterpretation.jl:1059
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1100
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1080
      From worker 2:    abstract_eval_statement at ./compiler/abstractinterpretation.jl:1220
      From worker 2:    typeinf_local at ./compiler/abstractinterpretation.jl:1542
      From worker 2:    typeinf_nocycle at ./compiler/abstractinterpretation.jl:1602
      From worker 2:    _typeinf at ./compiler/typeinfer.jl:214
      From worker 2:    typeinf at ./compiler/typeinfer.jl:209
      From worker 2:    typeinf_edge at ./compiler/typeinfer.jl:811
      From worker 2:    abstract_call_method at ./compiler/abstractinterpretation.jl:516
      From worker 2:    abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:142
      From worker 2:    abstract_call_known at ./compiler/abstractinterpretation.jl:1059
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1100
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1080
      From worker 2:    abstract_eval_statement at ./compiler/abstractinterpretation.jl:1220
      From worker 2:    typeinf_local at ./compiler/abstractinterpretation.jl:1525
      From worker 2:    typeinf_nocycle at ./compiler/abstractinterpretation.jl:1602
      From worker 2:    _typeinf at ./compiler/typeinfer.jl:214
      From worker 2:    typeinf at ./compiler/typeinfer.jl:209
      From worker 2:    typeinf_edge at ./compiler/typeinfer.jl:811
      From worker 2:    abstract_call_method at ./compiler/abstractinterpretation.jl:516
      From worker 2:    abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:142
      From worker 2:    abstract_call_known at ./compiler/abstractinterpretation.jl:1059
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1100
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1080
      From worker 2:    abstract_eval_statement at ./compiler/abstractinterpretation.jl:1220
      From worker 2:    typeinf_local at ./compiler/abstractinterpretation.jl:1542
      From worker 2:    typeinf_nocycle at ./compiler/abstractinterpretation.jl:1602
      From worker 2:    _typeinf at ./compiler/typeinfer.jl:214
      From worker 2:    typeinf at ./compiler/typeinfer.jl:209
      From worker 2:    typeinf_edge at ./compiler/typeinfer.jl:811
      From worker 2:    abstract_call_method at ./compiler/abstractinterpretation.jl:516
      From worker 2:    abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:142
      From worker 2:    abstract_call_known at ./compiler/abstractinterpretation.jl:1059
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1100
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1080
      From worker 2:    abstract_eval_statement at ./compiler/abstractinterpretation.jl:1220
      From worker 2:    typeinf_local at ./compiler/abstractinterpretation.jl:1542
      From worker 2:    typeinf_nocycle at ./compiler/abstractinterpretation.jl:1602
      From worker 2:    _typeinf at ./compiler/typeinfer.jl:214
      From worker 2:    typeinf at ./compiler/typeinfer.jl:209
      From worker 2:    typeinf_edge at ./compiler/typeinfer.jl:811
      From worker 2:    abstract_call_method at ./compiler/abstractinterpretation.jl:516
      From worker 2:    abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:142
      From worker 2:    abstract_call_known at ./compiler/abstractinterpretation.jl:1059
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1100
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1080
      From worker 2:    abstract_eval_statement at ./compiler/abstractinterpretation.jl:1220
      From worker 2:    typeinf_local at ./compiler/abstractinterpretation.jl:1542
      From worker 2:    typeinf_nocycle at ./compiler/abstractinterpretation.jl:1602
      From worker 2:    _typeinf at ./compiler/typeinfer.jl:214
      From worker 2:    typeinf at ./compiler/typeinfer.jl:209
      From worker 2:    typeinf_edge at ./compiler/typeinfer.jl:811
      From worker 2:    abstract_call_method at ./compiler/abstractinterpretation.jl:516
      From worker 2:    abstract_call_gf_by_type at ./compiler/abstractinterpretation.jl:142
      From worker 2:    abstract_call_known at ./compiler/abstractinterpretation.jl:1059
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1100
      From worker 2:    abstract_apply at ./compiler/abstractinterpretation.jl:764
      From worker 2:    abstract_call_known at ./compiler/abstractinterpretation.jl:980
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1100
      From worker 2:    abstract_call at ./compiler/abstractinterpretation.jl:1080
      From worker 2:    abstract_eval_statement at ./compiler/abstractinterpretation.jl:1220
      From worker 2:    typeinf_local at ./compiler/abstractinterpretation.jl:1542
      From worker 2:    typeinf_nocycle at ./compiler/abstractinterpretation.jl:1602
      From worker 2:    _typeinf at ./compiler/typeinfer.jl:214
      From worker 2:    typeinf at ./compiler/typeinfer.jl:209
      From worker 2:    typeinf_ext at ./compiler/typeinfer.jl:897
      From worker 2:    typeinf_ext_toplevel at ./compiler/typeinfer.jl:930
      From worker 2:    typeinf_ext_toplevel at ./compiler/typeinfer.jl:926
      From worker 2:    jfptr_typeinf_ext_toplevel_10402 at /home/arakaki/repos/watch/julia/usr/lib/julia/sys.so (unknown line)
      From worker 2:    _jl_invoke at /home/arakaki/repos/watch/julia/src/gf.c:2242 [inlined]
      From worker 2:    jl_apply_generic at /home/arakaki/repos/watch/julia/src/gf.c:2424
      From worker 2:    jl_apply at /home/arakaki/repos/watch/julia/src/julia.h:1738 [inlined]
      From worker 2:    jl_type_infer at /home/arakaki/repos/watch/julia/src/gf.c:297
      From worker 2:    jl_generate_fptr at /home/arakaki/repos/watch/julia/src/jitlayers.cpp:340
      From worker 2:    jl_compile_method_internal at /home/arakaki/repos/watch/julia/src/gf.c:1975
      From worker 2:    jl_compile_method_internal at /home/arakaki/repos/watch/julia/src/gf.c:2241 [inlined]
      From worker 2:    _jl_invoke at /home/arakaki/repos/watch/julia/src/gf.c:2234 [inlined]
      From worker 2:    jl_apply_generic at /home/arakaki/repos/watch/julia/src/gf.c:2424
      From worker 2:    jl_apply at /home/arakaki/repos/watch/julia/src/julia.h:1738 [inlined]
      From worker 2:    start_task at /home/arakaki/repos/watch/julia/src/task.c:859
      From worker 2:
      From worker 2:    signal (11): Segmentation fault
      From worker 2:    in expression starting at none:0
      From worker 2:    jl_lookup_generic_ at /home/arakaki/repos/watch/julia/src/gf.c:2353 [inlined]
      From worker 2:    jl_apply_generic at /home/arakaki/repos/watch/julia/src/gf.c:2420
      From worker 2:    next at /home/arakaki/.julia/packages/Transducers/XkqLO/src/combinators.jl:261
      From worker 2:    _jl_invoke at /home/arakaki/repos/watch/julia/src/gf.c:2242 [inlined]
      From worker 2:    jl_apply_generic at /home/arakaki/repos/watch/julia/src/gf.c:2424
      From worker 2:    next at /home/arakaki/.julia/packages/Transducers/XkqLO/src/core.jl:289
      From worker 2:    _jl_invoke at /home/arakaki/repos/watch/julia/src/gf.c:2242 [inlined]
      From worker 2:    jl_apply_generic at /home/arakaki/repos/watch/julia/src/gf.c:2424
      From worker 2:    next at /home/arakaki/.julia/packages/Transducers/XkqLO/src/library.jl:54
      From worker 2:    _jl_invoke at /home/arakaki/repos/watch/julia/src/gf.c:2242 [inlined]
      From worker 2:    jl_apply_generic at /home/arakaki/repos/watch/julia/src/gf.c:2424
      From worker 2:    next at /home/arakaki/.julia/packages/Transducers/XkqLO/src/core.jl:748
      From worker 2:    _jl_invoke at /home/arakaki/repos/watch/julia/src/gf.c:2242 [inlined]
      From worker 2:    jl_apply_generic at /home/arakaki/repos/watch/julia/src/gf.c:2424
      From worker 2:    macro expansion at /home/arakaki/.julia/packages/Transducers/XkqLO/src/core.jl:181 [inlined]
      From worker 2:    __foldl__ at /home/arakaki/.julia/packages/Transducers/XkqLO/src/processes.jl:238
      From worker 2:    unknown function (ip: 0x7f7525b55b13)
      From worker 2:    _jl_invoke at /home/arakaki/repos/watch/julia/src/gf.c:2242 [inlined]
      From worker 2:    jl_apply_generic at /home/arakaki/repos/watch/julia/src/gf.c:2424
      From worker 2:    foldl_nocomplete at /home/arakaki/.julia/packages/Transducers/XkqLO/src/processes.jl:356
      From worker 2:    _jl_invoke at /home/arakaki/repos/watch/julia/src/gf.c:2242 [inlined]
      From worker 2:    jl_apply_generic at /home/arakaki/repos/watch/julia/src/gf.c:2424
      From worker 2:    _reduce_basecase at /home/arakaki/.julia/packages/FoldsDagger/AeMjL/src/FoldsDagger.jl:87
      From worker 2:    _jl_invoke at /home/arakaki/repos/watch/julia/src/gf.c:2242 [inlined]
      From worker 2:    jl_apply_generic at /home/arakaki/repos/watch/julia/src/gf.c:2424
      From worker 2:    _reduce_basecase_zip at /home/arakaki/.julia/packages/FoldsDagger/AeMjL/src/darray.jl:49
      From worker 2:    _jl_invoke at /home/arakaki/repos/watch/julia/src/gf.c:2242 [inlined]
      From worker 2:    jl_apply_generic at /home/arakaki/repos/watch/julia/src/gf.c:2424
      From worker 2:    jl_apply at /home/arakaki/repos/watch/julia/src/julia.h:1738 [inlined]
      From worker 2:    do_apply at /home/arakaki/repos/watch/julia/src/builtins.c:688
      From worker 2:    macro expansion at /home/arakaki/.julia/packages/Dagger/Kcq9h/src/processor.jl:187 [inlined]
      From worker 2:    #49 at ./threadingconstructs.jl:170
      From worker 2:    _jl_invoke at /home/arakaki/repos/watch/julia/src/gf.c:2242 [inlined]
      From worker 2:    jl_apply_generic at /home/arakaki/repos/watch/julia/src/gf.c:2424
      From worker 2:    jl_apply at /home/arakaki/repos/watch/julia/src/julia.h:1738 [inlined]
      From worker 2:    start_task at /home/arakaki/repos/watch/julia/src/task.c:859
      From worker 2:    Allocations: 24982281 (Pool: 24975334; Big: 6947); GC: 31
Worker 2 terminated.┌ Warning: Worker 2 died on thunk 9, rescheduling work
└ @ Dagger.Sch ~/.julia/packages/Dagger/Kcq9h/src/sch/Sch.jl:217

ERROR: LoadError: AssertionError: No workers left for fault handling!
Stacktrace:
  [1] handle_fault(ctx::Context, state::Dagger.Sch.ComputeState, thunk::Thunk, oldproc::OSProc)
    @ Dagger.Sch ~/.julia/packages/Dagger/Kcq9h/src/sch/fault-handler.jl:112
  [2] (::Dagger.Sch.var"#53#59"{Context, Dagger.Sch.ComputeState, OSProc, Int64})()
    @ Dagger.Sch ~/.julia/packages/Dagger/Kcq9h/src/sch/Sch.jl:223
  [3] lock(f::Dagger.Sch.var"#53#59"{Context, Dagger.Sch.ComputeState, OSProc, Int64}, l::ReentrantLock)
    @ Base ./lock.jl:187
  [4] compute_dag(ctx::Context, d::Thunk; options::Nothing)
    @ Dagger.Sch ~/.julia/packages/Dagger/Kcq9h/src/sch/Sch.jl:222
  [5] compute(ctx::Context, d::Thunk; options::Nothing)
    @ Dagger ~/.julia/packages/Dagger/Kcq9h/src/compute.jl:31
  [6] collect(ctx::Context, t::Thunk; options::Nothing)
    @ Dagger ~/.julia/packages/Dagger/Kcq9h/src/compute.jl:8
  [7] #collect#85
    @ ~/.julia/packages/Dagger/Kcq9h/src/compute.jl:10 [inlined]
  [8] collect
    @ ~/.julia/packages/Dagger/Kcq9h/src/compute.jl:10 [inlined]
  [9] transduce_dagger(rf0::Function, init::Transducers.OnInit{var"#1#5"}, xs0::Dagger.DArray{Any, 1, typeof(cat)}; simd::Val{false}, basesize::Nothing)
    @ FoldsDagger ~/.julia/packages/FoldsDagger/AeMjL/src/FoldsDagger.jl:69
 [10] #transduce_dagger#3
    @ ~/.julia/packages/FoldsDagger/AeMjL/src/FoldsDagger.jl:48 [inlined]
 [11] transduce
    @ ~/.julia/packages/FoldsDagger/AeMjL/src/FoldsDagger.jl:122 [inlined]
 [12] transduce
    @ ~/.julia/packages/Transducers/XkqLO/src/executors.jl:62 [inlined]
 [13] _fold
    @ ~/.julia/packages/FLoops/GooJc/src/reduce.jl:361 [inlined]
 [14] _fold
    @ ~/.julia/packages/FLoops/GooJc/src/reduce.jl:359 [inlined]
 [15] macro expansion
    @ ~/.julia/packages/FLoops/GooJc/src/reduce.jl:341 [inlined]
 [16] gnorm!(mixer::typeof(ampnorm_mixer), ys::Dagger.DArray{Any, 1, typeof(cat)}, xs::Dagger.DArray{Any, 1, typeof(cat)}, ex::Nothing)
    @ Main ~/tmp/mwe/mwe.jl:25
 [17] gnorm!(mixer::typeof(ampnorm_mixer), ys::Dagger.DArray{Any, 1, typeof(cat)}, xs::Dagger.DArray{Any, 1, typeof(cat)})
    @ Main ~/tmp/mwe/mwe.jl:24
 [18] top-level scope
    @ ~/tmp/mwe/mwe.jl:45
 [19] include(fname::String)
    @ Base.MainInclude ./client.jl:451
 [20] top-level scope
    @ none:1
in expression starting at /home/arakaki/tmp/mwe/mwe.jl:45
```

</details>

I don't _think_ I'm doing anything unsafe in my packages but I'm not sure where to start. Does this error look familiar to somebody?

--- a/stdlib/Serialization/src/Serialization.jl
+++ b/stdlib/Serialization/src/Serialization.jl
@@ -1289,6 +1289,8 @@ function deserialize_typename(s::AbstractSerializer, number)
                 tn.mt.kwsorter = kws
             end
         end
+    elseif makenew
+        tn.mt = Symbol.name.mt
     end
     return tn::Core.TypeName
 end

--- a/stdlib/Serialization/test/runtests.jl
+++ b/stdlib/Serialization/test/runtests.jl
@@ -609,3 +609,18 @@ let s = join(rand('a':'z', 1024)), io = IOBuffer()
     s2 = deserialize(io)
     @test Base.summarysize(s2) < 2*sizeof(s)
 end
+
+# issue #39895
+@eval Main begin
+    using Test, Serialization
+    let g = gensym(:g)
+        closure = eval(:(f -> $g(x) = f(x)))
+        inc(x) = x + 1
+        b = IOBuffer()
+        serialize(b, closure(inc))
+        seekstart(b)
+        f = deserialize(b)
+        # this should not crash
+        @test_broken f(1) == 2
+    end
+end

